### PR TITLE
OSL-148: Adding Redis cache backend & caching shareableListPublic query

### PR DESCRIPTION
**Pull request by katerinachinnappan**
## Goal

Adding Redis cache backend and caching response for `shareableListPublic` query. Previous PR had to be reverted, see [slack thread](https://pocket.slack.com/archives/C04HWMUKLN5/p1679368237398669). Removed `persistedQueries` from ApolloServer config as we only cache `shareableListPublic` query and that query is not big.

## Todos

- [ ] deploy to dev, ensure web-client works properly

## Tickets

- [https://getpocket.atlassian.net/browse/OSL-148](https://getpocket.atlassian.net/browse/OSL-148)

### CODE REVIEW

**Review by jpetto — CHANGES_REQUESTED** [review]
a few notes in code, but also:

i think we need to add tests around the cache. the result of caching _should_ be a `cache-control` header value sent in the response from calling the `shareableListPublic` query. our integration tests _do_ provide access to these headers, so theoretically we could verify that `cache-control` is set for `shareableListPublic` and _not_ set for all other queries. 

i don't know that we have prior art for this exact scenario. depending on my health levels tomorrow, we could pair or async investigate.

File: src/public/server.ts [hunk 1]
```diff
@@ -10,17 +10,28 @@ import {
...
 ): ApolloServer<IPublicContext> {
+  const cache = getRedisCache();
   const defaultPlugins = [
+    // On initialization, this plugin automatically begins caching responses according to field settings
+    // see shareableListPublic cache control settings
+    responseCachePlugin(),
     sentryPlugin,
     ApolloServerPluginDrainHttpServer({ httpServer }),
+    ApolloServerPluginCacheControl({
+      defaultMaxAge: config.app.defaultMaxAge,
```
[the remark this answers is not in the document]
Comment (katerinachinnappan) [reply src/public/server.ts:33]: @jpetto should we also default it to a minute then?

[the remark this answers is not in the document]
Comment (katerinachinnappan) [reply src/public/server.ts:35]: @jpetto defaulted to 0 as we don't want to cache any other queries by default.

File: schema-public.graphql [hunk 2]
```diff
@@ -46,6 +46,55 @@ type ShareableList implements ShareableListInterface {
   listItems: [ShareableListItem!]!
 }
 
+"""
+A user-created list of Pocket saves that can be shared publicly.
+"""
+type ShareableListPublic implements ShareableListInterface
```
Comment (jpetto) [inline schema-public.graphql:52]: nicely done creating another type we can explicitly cache.

File: src/public/resolvers/queries/ShareableList.ts [hunk 3]
```diff
@@ -43,14 +44,20 @@ export async function getShareableList(
...
   if (!list) {
     throw new NotFoundError(externalId);
   }
-
+  // dynamically providing caching controll in the resolver
+  // PUBLIC scope because response is accessible to public not a single user
+  info.cacheControl.setCacheHint({
+    maxAge: config.redis.querySettings.shareableListPublic.maxAge, // The maximum amount of time the field's cached value is valid, in seconds.
+    scope: config.redis.querySettings.shareableListPublic.scope,
+  });
```
Comment (jpetto) [inline src/public/resolvers/queries/ShareableList.ts:60]: with the `@cacheControl` directives, i'm pretty sure we don't need to specify caching here as well. we should be able to remove this part.

**Review by jpetto — CHANGES_REQUESTED** [review]
great job re-working this PR! a few small things to clean up, but the solution works!

i might also ask - just to be sure, could we add `cache-control` checks in the mutation integration tests as well?

**Review by jpetto — APPROVED** [review]
i think this is ready to go! give the team a heads up in the project channel before merge - but then send it.

great work!

**Pull request merged**